From f4dc4f5e627cf743f1eb9e96ccf00d270df0eebc Mon Sep 17 00:00:00 2001
From: Adhemerval Zanella <adhemerval.zanella@linaro.org>
Date: Thu, 21 Jul 2022 14:12:01 -0300
Subject: [PATCH 023/142] argp: Expand argp_usage, _option_is_short, and
 _option_is_end

The argp code uses some clever macro redefine to avoid need to
duplicate the optimized static inline implementations for
argp_usage, _option_is_short, and _option_is_end.  This however
leads to some build issues with clang, since some function
prototypes are redefined to add the hidden attribute with
libc_hidden_proto.

To avoid extensive changes on internal headers, just expand the
function implementations and avoid the macro redefine tricks.
---
 argp/argp-xinl.c | 34 +++++++++++++++++++++++-----------
 argp/argp.h      | 10 +++-------
 2 files changed, 26 insertions(+), 18 deletions(-)

diff --git a/argp/argp-xinl.c b/argp/argp-xinl.c
index 4bd4671699..e9d654499f 100644
--- a/argp/argp-xinl.c
+++ b/argp/argp-xinl.c
@@ -25,19 +25,31 @@
 # include <features.h>
 #endif
 
-#ifndef __USE_EXTERN_INLINES
-# define __USE_EXTERN_INLINES	1
-#endif
-#define ARGP_EI
-#undef __OPTIMIZE__
-#define __OPTIMIZE__ 1
 #include <argp.h>
 
-/* Add weak aliases.  */
-#if _LIBC - 0 && defined (weak_alias)
-
+void
+__argp_usage (const struct argp_state *__state)
+{
+  __argp_state_help (__state, stderr, ARGP_HELP_STD_USAGE);
+}
 weak_alias (__argp_usage, argp_usage)
+
+int
+__option_is_short (const struct argp_option *__opt)
+{
+  if (__opt->flags & OPTION_DOC)
+    return 0;
+  else
+    {
+      int __key = __opt->key;
+      return __key > 0 && __key <= UCHAR_MAX && isprint (__key);
+    }
+}
 weak_alias (__option_is_short, _option_is_short)
+
+int
+__option_is_end (const struct argp_option *__opt)
+{
+  return !__opt->key && !__opt->name && !__opt->doc && !__opt->group;
+}
 weak_alias (__option_is_end, _option_is_end)
-
-#endif
diff --git a/argp/argp.h b/argp/argp.h
index efae2a6587..99d3b4d788 100644
--- a/argp/argp.h
+++ b/argp/argp.h
@@ -518,17 +518,13 @@ extern void *__argp_input (const struct argp *__restrict __argp,
 #  define __option_is_end _option_is_end
 # endif
 
-# ifndef ARGP_EI
-#  define ARGP_EI __extern_inline
-# endif
-
-ARGP_EI void
+__extern_inline void
 __argp_usage (const struct argp_state *__state)
 {
   __argp_state_help (__state, stderr, ARGP_HELP_STD_USAGE);
 }
 
-ARGP_EI int
+__extern_inline int
 __NTH (__option_is_short (const struct argp_option *__opt))
 {
   if (__opt->flags & OPTION_DOC)
@@ -540,7 +536,7 @@ __NTH (__option_is_short (const struct argp_option *__opt))
     }
 }
 
-ARGP_EI int
+__extern_inline int
 __NTH (__option_is_end (const struct argp_option *__opt))
 {
   return !__opt->key && !__opt->name && !__opt->doc && !__opt->group;
-- 
2.34.1

